signatureverifier/src/main/kotlin/io/github/rakutentech/signatureverifier/verification/PublicKeyCache.kt

```diff
@@ -1,45 +1,84 @@
+import com.google.gson.reflect.TypeToken
+import io.github.rakutentech.signatureverifier.SignatureVerifier
 import io.github.rakutentech.signatureverifier.api.PublicKeyFetcher
-import java.io.IOException
+import java.io.File
 
 internal class PublicKeyCache(
     private val keyFetcher: PublicKeyFetcher,
     context: Context,
-    prefEncryptor: SharedPreferenceEncryptor? = null
+    baseUrl: String,
+    encryptor: AesEncryptor? = null,
+    testKeys: MutableMap<String, String>? = null
 ) {
 
-    private val encryptor: SharedPreferenceEncryptor? by lazy {
-        prefEncryptor ?: try {
-            SharedPreferenceEncryptor(context)
-        } catch (ex: SignatureVerifierException) {
-            Log.d("SignatureVerifier", ex.message, ex.cause)
-            null
-        }
+    private val encryptor: AesEncryptor by lazy { encryptor ?: AesEncryptor() }
+
+    private val file: File by lazy {
+        // replace all non-alphanumeric characters to '.':
+        // - multiple/group of non-alphanumeric will be replace by one '.'
+        // - trailing period is removed
+        val filepath = baseUrl.replace(Regex("[^a-zA-Z0-9]+"), ".").replace(Regex(".$"), "")
+        File(
+            context.noBackupFilesDir,
+            ("signature.keys.$filepath").take(MAX_PATH)
+        )
     }
 
-    operator fun get(keyId: String): String? {
-        if (encryptor == null) {
-            return null
-        }
+    @VisibleForTesting
+    internal val keys: MutableMap<String, String> by lazy {
+        testKeys
+            ?: if (file.exists()) {
+                val text = file.readText()
+
+                if (text.isNotBlank()) {
+                    val type = object : TypeToken<MutableMap<String, String>>() {}.type
+                    Gson().fromJson(text, type)
+                } else {
+                    mutableMapOf()
+                }
+            } else {
+                mutableMapOf()
+            }
+    }
+
+    operator fun get(keyId: String): String {
+        val encryptedKey = keys[keyId]
 
-        return try {
-            val id = encryptor?.getEncryptedKey(keyId) ?: keyFetcher.fetch(keyId)
-            encryptor?.putEncryptedKey(keyId, id)
-            id
-        } catch (e: IOException) {
-            Log.e(TAG, "Failed to fetch public key", e)
-            null
+        return if (encryptedKey != null) {
+            encryptor.decrypt(encryptedKey) ?: fetch(keyId)
+        } else {
+            fetch(keyId)
         }
     }
 
     fun remove(keyId: String) {
-        encryptor?.putEncryptedKey(keyId, null)
+        keys.remove(keyId)
+
+        file.writeText(Gson().toJson(keys))
+    }
+
+    @SuppressWarnings("TooGenericExceptionCaught")
+    private fun fetch(keyId: String): String {
+        val key = try {
+            keyFetcher.fetch(keyId)
+        } catch (ex: Exception) {
+            SignatureVerifier.callback?.let { it(ex) }
+            ""
+        }
+        encryptor.encrypt(key)?.let {
```

Comment:
In case of fetch error, why do we write empty ("") key to the store?

Reply:
The encryption should fail and return null.
But changed this implementation to make key null when fetch throws exception.